File: IntroOS/Presentation/sample_code/synchronization_primitives.java
class Counter {
    private int counter = 0;

    public synchronized void increment() {
        counter++;
        System.out.println("Counter: " + counter);
    }
}

public class synchronization_primitives {
    public static void main(String[] args) {
        Counter counter = new Counter();

        Thread thread1 = new Thread(() -> {
            counter.increment();
        });

        Thread thread2 = new Thread(() -> {
            counter.increment();
        });

        thread1.start();
        thread2.start();

        try {
            thread1.join();
            thread2.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: IntroOS/Presentation/sample_code/synchronization_primitives.py
import threading

mutex = threading.Lock()
counter = 0


def increment():
    global counter
    mutex.acquire()
    counter += 1
    print("Counter:", counter)
    mutex.release()


thread1 = threading.Thread(target=increment)
thread2 = threading.Thread(target=increment)

thread1.start()
thread2.start()

thread1.join()
thread2.join()
